### PR TITLE
Add extract_metadata_from_bazel_xml.py explainers

### PR DESCRIPTION
**Pull request by jtattermusch**
- improve comments to make the entire extraction process more understandable
- a slight refactoring to improve readability

Also adds some TODOs to better document how the bazel -> build.yaml extraction process can be simplified in the future.

### CODE REVIEW

File: tools/buildgen/extract_metadata_from_bazel_xml.py [hunk 1]
```diff
@@ -567,6 +595,16 @@ def _generate_build_extra_metadata_for_tests(tests, bazel_rules):
...
+    """Try detecting some unusual situations and warn about them."""
+    for tgt in build_yaml_like['targets']:
+        if tgt['build'] == 'test':
+            for src in tgt['src']:
+                if src.startswith('src/') and not src.endswith('.proto'):
+                    print('source file from under "src/" tree used in test ' +
+                          tgt['name'] + ': ' + src)
+
+
 # extra metadata that will be used to construct build.yaml
```
Comment (markdroth) [inline tools/buildgen/extract_metadata_from_bazel_xml.py:608]: It seems strange to hard-code all of this data inside of this script instead of putting it in an external data file somewhere.  Could this be moved to build_handwritten.yaml, or something similar?

File: tools/buildgen/extract_metadata_from_bazel_xml.py [hunk 2]
```diff
@@ -410,8 +437,8 @@ def _extract_cc_tests(bazel_rules):
     return list(sorted(result))
 
 
-def _filter_cc_tests(tests):
-    """Filters out tests that we don't want or we cannot build them reasonably"""
+def _exclude_unwanted_cc_tests(tests):
+    """Filters out bazel tests that we don't want to run with other build systems or we cannot build them reasonably"""
 
     # most qps tests are autogenerated, we are fine without them
```
Comment (markdroth) [inline tools/buildgen/extract_metadata_from_bazel_xml.py:443]: It would be nice to move the specific cases we're excluding here out of this file and into a data file, so that the specific inputs can be modified without changing this script.

Alternatively, why do we still need any of the C/C++ tests to run in non-bazel builds?  Couldn't we just ignore all tests here?

File: tools/buildgen/extract_metadata_from_bazel_xml.py [hunk 3]
```diff
@@ -937,31 +975,86 @@ def _generate_build_extra_metadata_for_tests(tests, bazel_rules):
...
-all_targets_dict = _generate_build_metadata(all_metadata, bazel_rules)
+# Step 2: Extract the known bazel cc_test tests. While most tests
+# will be buildable with other build systems just fine, some of these tests
+# would be too difficult to build and run with other build systems,
+# so we simply the ones we don't want.
+tests = _exclude_unwanted_cc_tests(_extract_cc_tests(bazel_rules))
+
+# Step 3: Generate the "extra metadata" for all our build targets.
+# While the bazel rules give us most of the information we need,
+# the legacy "build.yaml" format requires some additional fields that
```
Comment (markdroth) [inline tools/buildgen/extract_metadata_from_bazel_xml.py:1000]: Can you list the specific extra fields that need to be populated and what each of them indicates?

File: tools/buildgen/extract_metadata_from_bazel_xml.py [hunk 4]
```diff
@@ -937,31 +975,86 @@ def _generate_build_extra_metadata_for_tests(tests, bazel_rules):
     'deps("//src/proto/...")',
 ]
 
+# Step 1: run a bunch of "bazel query --output xml" queries to collect
+# the raw build metadata from the bazel build.
+# At the end of this step we will have a dictionary of bazel rules
+# that are interesting to us (libraries, binaries, etc.) along
+# with their most important metadata (sources, headers, dependencies)
```
Comment (markdroth) [inline tools/buildgen/extract_metadata_from_bazel_xml.py:982]: Can you show an example of what the resulting rule dict will look like?

File: tools/buildgen/extract_metadata_from_bazel_xml.py [hunk 5]
```diff
@@ -937,31 +975,86 @@ def _generate_build_extra_metadata_for_tests(tests, bazel_rules):
...
 bazel_rules = {}
 for query in _BAZEL_DEPS_QUERIES:
     bazel_rules.update(
         _extract_rules_from_bazel_xml(_bazel_query_xml_tree(query)))
 
+# Step 1a: Knowing the transitive closure of dependencies will make
+# the postprocessing simpler, so compute the info for all our rules.
```
Comment (markdroth) [inline tools/buildgen/extract_metadata_from_bazel_xml.py:989]: I assume this step modifies `bazel_rules` in place.  Can you show an example of how a given rule might be modified, so a reader can tell what the data should look like at this point?

File: tools/buildgen/extract_metadata_from_bazel_xml.py [hunk 6]
```diff
@@ -937,31 +975,86 @@ def _generate_build_extra_metadata_for_tests(tests, bazel_rules):
...
+all_extra_metadata.update(
+    _generate_build_extra_metadata_for_tests(tests, bazel_rules))
+
+# Step 4: Generate the final metadata for all the targets.
+# This is done by combining the bazel build metadata and the "extra metadata"
+# we obtained in the previous step.
+# In this step, we also perform some interesting massaging of the target metadata
+# to end up with a result that is as similar to the legacy build.yaml data
+# as possible.
+# - Some targets get renamed (to match the legacy build.yaml target names)
```
Comment (markdroth) [inline tools/buildgen/extract_metadata_from_bazel_xml.py:1036]: Why is this needed?  Can't we name these whatever we want?

File: tools/buildgen/extract_metadata_from_bazel_xml.py [hunk 7]
```diff
@@ -937,31 +975,86 @@ def _generate_build_extra_metadata_for_tests(tests, bazel_rules):
...
+# - Some intermediate libraries get elided ("expanded") to better match the set
+#   of targets provided by the legacy build.yaml build
+all_targets_dict = _generate_build_metadata(all_extra_metadata, bazel_rules)
```
Comment (markdroth) [inline tools/buildgen/extract_metadata_from_bazel_xml.py:1039]: Can you show an example of what `all_targets_dict` looks like after this step?

File: tools/buildgen/extract_metadata_from_bazel_xml.py [hunk 8]
```diff
@@ -937,31 +975,86 @@ def _generate_build_extra_metadata_for_tests(tests, bazel_rules):
...
+
+# Step 5: convert the dictionary with all the targets to a dict that has
```
Comment (markdroth) [inline tools/buildgen/extract_metadata_from_bazel_xml.py:1041]: Can you show an example of what `build_yaml_like` looks like after this step?

File: tools/buildgen/extract_metadata_from_bazel_xml.py [hunk 9]
```diff
@@ -937,31 +975,86 @@ def _generate_build_extra_metadata_for_tests(tests, bazel_rules):
...
 build_yaml_like = _convert_to_build_yaml_like(all_targets_dict)
 
-# if a test uses source files from src/ directly, it's a little bit suspicious
-for tgt in build_yaml_like['targets']:
-    if tgt['build'] == 'test':
-        for src in tgt['src']:
-            if src.startswith('src/') and not src.endswith('.proto'):
-                print('source file from under "src/" tree used in test ' +
-                      tgt['name'] + ': ' + src)
+_detect_and_print_issues(
```
Comment (markdroth) [inline tools/buildgen/extract_metadata_from_bazel_xml.py:1050]: Please write this as:

```
# detect and report some suspicious situations we've seen before
_detect_and_print_issues(build_yaml_like)
```

File: tools/buildgen/extract_metadata_from_bazel_xml.py (hunk 6 above)
Comment (jtattermusch) [reply tools/buildgen/extract_metadata_from_bazel_xml.py:1036]: Basically I think it's these reasons:
- in the original PR I wanted to avoid changing too many things at the same time and avoid people getting confused by some well know targets suddenly being missing (the PR was already a big conceptual change)
- Makefile/cmake and also language-specific generators rely on some build targets being called exactly as they are.
- I wanted to test to have similar names as before (otherwise they would be called like `test_cpp_TEST_NAME` because some bazel targets are not top-level.
- many of our other scripts invoke executables (e.g. "qps_json_driver") by their name. Without renaming, the target binary name would have been `test_cpp_qps_qps_json_driver` which is both ugly and would also have broken our tests.

The renaming is something that can be revisited in the future.

File: tools/buildgen/extract_metadata_from_bazel_xml.py (hunk 2 above)
Comment (jtattermusch) [reply tools/buildgen/extract_metadata_from_bazel_xml.py:443]: In the long run, we probably can remove some C/C++ tests in  non-bazel builds but there are some concerns:

- if the Make/cmake builds don't have any tests at all, how can we be sure what they've build actually works? (so we need at least some "smoke tests"). Btw, the build flags between bazel / non-bazel flag might differ (sometimes it's for interesting reasons that are not easy to overcome) which makes it even more important to have at least some tests for make/cmake.

- our current portability suite actually runs make/cmake tests and we'd need to migrate it to use bazel first (which brings the question whether it's possible to get a good enough coverage of different compilers / distros etc, with bazel - it's something we haven't really tried yet).

- some things that are considered "tests" in build.yaml based builds are actually binaries you'd want to be able to build anyway (qps_json_worker, interop_client, interop_server, grpc_cli) so it's unclear how much make/cmake simplification we would gain by removing just some (but not all) tests.

File: tools/buildgen/extract_metadata_from_bazel_xml.py (hunk 3 above)
Comment (jtattermusch) [reply tools/buildgen/extract_metadata_from_bazel_xml.py:1000]: The fields being set and explanation for some of them is here:
https://github.com/grpc/grpc/blob/9e095b2076e1df72d2241128b9c6714d4cc8264b/tools/buildgen/extract_metadata_from_bazel_xml.py#L575

General documentation of "build.yaml" contents is here:
https://github.com/grpc/grpc/blob/master/templates/README.md
That is good enough for general understanding what the fields mean, but the docs might be out of date or incorrect in some cases (build.yaml has been with us from pretty much the project start). While some of the field name seem self-explanatory, I don't really have a way to know what exactly the fields are supposed to mean because they way the they were added to `build.yaml` was mostly random and they were added on ad-hoc and as-needed fashion. So when someone needed a new field to achieve something in our build system, they added a new attribute and the exact meaning was defined implicitly by implementing its behavior in our build file generators.

For the above reasons, there's actually no clearly defined list of extra fields that need to be populated - it's basically empirical (I basically just tried to match exactly what was in the original build.yaml because that was the only way to prevent things from exploding - there are too many special cases in our build generators to address all of them at once).

In the future, the right thing to do is to revisit the "extra metadata" fields being used in _BUILD_EXTRA_METADATA and get rid of all of them one by one  (either by removing the need for a field entirely or somehow adding the information to the bazel rule so we can consume it) - unlike before, this is now possible because bazel build is now our source of truth and the big bazel related changes have already settled down (and most builds are mostly green).

File: tools/buildgen/extract_metadata_from_bazel_xml.py [hunk 10]
```diff
@@ -13,6 +13,15 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+# Script to extract build metadata from bazel BUILD.
+# To avoid having two sources of truth for the build metadata (build
+# targets, source files, header files etc.), this script analyzes the contents
+# of bazel BUILD files and generates a YAML file (currently called
+# build_autogenerated.yaml). The format and semantics of the generated YAML files
+# is chosen to match the format of a "build.yaml" file, which used
```
[the remark this answers is not in the document]
Comment (jtattermusch) [reply tools/buildgen/extract_metadata_from_bazel_xml.py:21]: General documentation of "build.yaml" contents is here:
https://github.com/grpc/grpc/blob/master/templates/README.md, but as noted in another comment, there is really no spec and meaning of some of the constructs is defined empirically.
I do not plan to provide better documentation for the legacy build.yaml format as the plan is to simplify build.yaml significantly or get rid of it entirely in the future.

File: tools/buildgen/extract_metadata_from_bazel_xml.py (hunk 7 above)
Comment (jtattermusch) [reply tools/buildgen/extract_metadata_from_bazel_xml.py:1067]: done.

File: tools/buildgen/extract_metadata_from_bazel_xml.py (hunk 8 above)
Comment (jtattermusch) [reply tools/buildgen/extract_metadata_from_bazel_xml.py:1069]: done

File: tools/buildgen/extract_metadata_from_bazel_xml.py (hunk 9 above)
Comment (jtattermusch) [reply tools/buildgen/extract_metadata_from_bazel_xml.py]: done.

File: tools/buildgen/extract_metadata_from_bazel_xml.py (hunk 4 above)
Comment (jtattermusch) [reply tools/buildgen/extract_metadata_from_bazel_xml.py:991]: done.

File: tools/buildgen/extract_metadata_from_bazel_xml.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (markdroth) [reply tools/buildgen/extract_metadata_from_bazel_xml.py:21]: Okay.  In that case, please add a TODO indicating that we plan to eliminate use of this legacy format.

File: tools/buildgen/extract_metadata_from_bazel_xml.py (hunk 6 above)
Comment (markdroth) [reply tools/buildgen/extract_metadata_from_bazel_xml.py:1056]: Okay.  Please document these reasons here, with a TODO to clean this up.

File: tools/buildgen/extract_metadata_from_bazel_xml.py (hunk 2 above)
Comment (markdroth) [reply tools/buildgen/extract_metadata_from_bazel_xml.py:452]: Please document these concerns in TODO comments.

File: tools/buildgen/extract_metadata_from_bazel_xml.py (hunk 1 above)
Comment (markdroth) [reply tools/buildgen/extract_metadata_from_bazel_xml.py:617]: I didn't see a response to this question.

File: tools/buildgen/extract_metadata_from_bazel_xml.py (hunk 6 above)
Comment (jtattermusch) [reply tools/buildgen/extract_metadata_from_bazel_xml.py:1075]: Documented the reasons in a comment and added a TODO.

File: tools/buildgen/extract_metadata_from_bazel_xml.py (hunk 2 above)
Comment (jtattermusch) [reply tools/buildgen/extract_metadata_from_bazel_xml.py:455]: Documented the reasons/concerts in a comment at "Step 2" (extracting tests).

**Comment by jtattermusch** [comment]
@markdroth  are there any more comments? I'd like to get this in sooner than later.

**Review by markdroth — APPROVED** [review]

**Comment by markdroth** [comment]
Thanks for doing this!
